### PR TITLE
Add a program to enforce ignore pattern

### PR DESCRIPTION
**Pull request by louischan-oursky**
This program implements https://github.com/oursky/devsecops-ci/issues/56

This PR just introduces such program. It does not adhere to the current make-based build system nor support git commit range.

### CODE REVIEW

File: src/ignorefile/Makefile [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+test:
+	dune runtest
+
+.PHONY: build
+build:
+	dune build cmd/ignorefile.exe
+
+.PHONY: clean
+clean:
+	dune clean
```
Comment (hidden-dimensions) [inline src/ignorefile/Makefile]: The Makefile should be
```
.PHONY: all build check 

all: build
	@:

build:
	@apk add opam
	@opam install . --deps-only
	@opam config exec make build

check: FORCE
	@cmd/ignorefile.exe --conf-file="$(CONF_FILE)" --target-dir="$(TARGET_DIR)" --verbose="$(VERBOSE)" --commit-range="$(COMMIT_RANGE)"

FORCE:
```

You may need to change the `check` command.

File: src/ignorefile/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,24 @@
+FROM ocaml/opam2:alpine as builder
```
Comment (hidden-dimensions) [inline src/ignorefile/Dockerfile]: Currently everything should put into top level Dockerfile, that Dockerfile will call `make build` on your source folder.
You may need some magic there for multi-stage build.

File: src/ignorefile/Makefile (hunk 1 above)
Comment (louischan-oursky) [reply src/ignorefile/Makefile]: I have been thinking how should we support config file and commit range. I think it is best for a entry program to parse config and handle commit range, and then invoke the individual checker command with correct options and arguments. The advantage is that we can easily introduce new checker.

File: src/ignorefile/Dockerfile (hunk 2 above)
Comment (louischan-oursky) [reply src/ignorefile/Dockerfile]: Multi-stage build with make is quite tricky. It is much easier if we can let each checker to build their own image. Say the docker image should have an self-contained executable at `/checker`. In the outermost dockerfile, we just have to use multi-stage build to copy all executables to build a final image. The problem is about how to package python program as a self-contained executable. This tool seems promising https://github.com/facebookincubator/xar Maybe we should refactor the current build architecture before adding new checkers.

File: src/ignorefile/Makefile (hunk 1 above)
Comment (hidden-dimensions) [reply src/ignorefile/Makefile]: But some check is not about commit (e.g. you may still want to run `npm audit` even it's not within the commit?) Then we need to introduce 2 type of checkers (one for commits, one for current checkout files).

File: src/ignorefile/Dockerfile (hunk 2 above)
Comment (hidden-dimensions) [reply src/ignorefile/Dockerfile]: Just copy the binary might not work unless you do all static link.

**Pull request closed without merging**